### DISCUSSION TITLE
How to avoid empty commit?

### DISCUSSION THREAD

**Discussion by SunilDSK** [alirezanet/Husky.Net#88] [Q&A]
### Version

0.6.0

### Details

I have a webapi project with environment specific appsettings.json file. I don't want the `appsettings.Development.json` to be committed or pushed(its already tracked). I have a pre-commit hook with one command - `git reset appsettings.Development.json`. When I have `appsettings.Development.json` as only file, the pre-commit is unstaging the `appsettings.Development.json` file, but it is also creating an empty commit with no files. How do I avoid the empty commit?

### Steps to reproduce

- Step 1: Once dotnet husky is installed, run the command - `$ dotnet husky add pre-commit -c "git reset appsettings.Development.json"`.
- Step 2: Then modify the file `appsettings.Development.json`.
- Step 3: Then stage the file and try to commit it
- Step 4: The git will unstage the `appsettings.Development.json` file and create an empty commit.

**Comment by alirezanet** [comment]
Hi @SunilDSK,
I don't see it as a bug  although I didn't fully understand what do you expect Husky does for you?

if you wanna prevent commits if `appsettings.Development.json` is staged I would suggest to use the `task-runner` instead.

something like:
e.g pre-commit file
``` bash
#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

dotnet husky run -v --group "pre-commit"
```

task-runner.json

``` json
{
 "tasks": [
      {
         "name": "Verify staged files",
         "group": "pre-commit",
         "include": "appsettings.Development.json",
         "command": "cmd",
         "args": ["/c", "git reset appsettings.Development.json & exit 1", "${staged}"]
      }
]
}
```

This would automatically unstage the `appsettings.Development.json` file if it is staged and prevent the commits including this file.

also using taks-runner makes your console log prettier :)

**Reply by alirezanet** [reply]
@SunilDSK, I converted this issue to a discussion... so feel free to ask if you have other questions.

**Reply by SunilDSK** [reply]
@alirezanet, Thanks for the quick response 🥇. Unfortunately, the above task did your work. It didn't even commit the files when `appsettings.Development.json` was not modified. However, I did some digging. I was able to solve this issue with a shell script in pre-commit file- 💯 
```shell
#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

git reset appsettings.Development.json

if [ -n "$(git diff --cached --exit-code)" ]; then
  echo "there are changes";
else
  echo "no changes";
  exit 1
fi
exit 0
```
Can you please help me getting it working on task-runner.json? I failed trying these tasks, these did not work 👎🏽  - 
```json
{
   "tasks": [
      {
         "name": "Skip dev settings",
         "group": "pre-commit",
         "command": "cmd",
         "args": ["/c", "git reset appsettings.Development.json"]
      },
      {
         "name": "Verify if nothing is staged",
         "group": "pre-commit",
         "command": "cmd",
         "args": ["/c", "git diff","${staged}"]
      }
   ]
}
```

**Reply by alirezanet** [reply]
That's weird, Yesterday I tested the task I sent to you and it was working as expected! did you update also your pre-commit hook contents?

```
dotnet husky run --group "pre-commit"
```

usually, I don't recommend combining task-runner tasks and custom logic in the hooks, especially when you want to run git-related commands like (git reset). 

```
         "name": "Verify staged files", // name of the task (you can change it to whatever you want)
         "group": "pre-commit", // this group is needed if you want to run only certain tasks in certain hooks (name can be different, just make sure you also update it in the hook -> dotnet husky run -v --group "pre-commit")
         "include": ["appsettings.Development.json"], // include checks if appsettings.Development.json file is staged or not, if it is staged it will execute the task
         "command": "cmd",
         "args": ["/c", "git reset appsettings.Development.json & exit 1", "${staged}"] // using cmd it will reset the target file and exit with non-zero code to prevent commits
```

checking the task I noticed `include`  might be wrong, try to use it this way, (if it is on a different path, change it accordingly)
```
  "include": ["appsettings.Development.json"],
```

**Reply by alirezanet** [reply]
let me know if it didn't work, I'll try to check it again later today.
